packages/x-data-grid-premium/src/hooks/features/pivoting/useGridPivoting.ts

```diff
@@ -0,0 +1,403 @@
+              aggFunc,
+              hidden,
+            });
+            newModel.values = newSectionArray as GridPivotModel['values'];
+          } else if (targetSection === 'columns') {
+            const sort = isSameSection
+              ? prev.columns.find((item) => item.field === field)?.sort
+              : undefined;
+            newSectionArray.splice(toIndex, 0, { field, sort, hidden });
+            newModel.columns = newSectionArray as GridPivotModel['columns'];
+          } else if (targetSection === 'rows') {
+            newSectionArray.splice(toIndex, 0, { field, hidden });
+            newModel.rows = newSectionArray as GridPivotModel['rows'];
+          }
+        }
+        if (!isSameSection && originSection) {
+          (newModel[originSection] as (typeof prev)[typeof originSection]) = prev[
+            originSection
+          ].filter((f) => f.field !== field);
+        }
+        return newModel;
+      });
+    },
+    [apiRef, props.aggregationFunctions],
+  );
+
+  const setPivotActive = React.useCallback<GridPivotingApi['setPivotActive']>(
+    (callback) => {
+      if (!isPivotingAvailable) {
+        return;
+      }
+      apiRef.current.setState((state) => {
+        const newPivotMode =
+          typeof callback === 'function' ? callback(state.pivoting?.active) : callback;
+
+        if (state.pivoting?.active === newPivotMode) {
+          return state;
+        }
+
+        if (newPivotMode) {
+          nonPivotDataRef.current = getInitialData();
+        } else {
+          nonPivotDataRef.current = undefined;
+        }
+
+        const newPivotingState: GridPivotingState = {
+          ...state.pivoting,
+          ...computePivotingState({
+            ...state.pivoting,
+            active: newPivotMode,
+          }),
+          active: newPivotMode,
+        };
+
+        const newState = {
+          ...state,
+          pivoting: newPivotingState,
+        };
+        return newState;
+      });
+    },
+    [apiRef, computePivotingState, getInitialData, isPivotingAvailable],
+  );
+
+  const setPivotPanelOpen = React.useCallback<GridPivotingApi['setPivotPanelOpen']>(
+    (callback) => {
+      if (!isPivotingAvailable) {
+        return;
+      }
+      apiRef.current.setState((state) => ({
+        ...state,
+        pivoting: {
+          ...state.pivoting,
+          panelOpen:
+            typeof callback === 'function' ? callback(state.pivoting?.panelOpen) : callback,
+        },
+      }));
+    },
+    [apiRef, isPivotingAvailable],
+  );
+
+  const addColumnMenuButton = React.useCallback<GridPipeProcessor<'columnMenu'>>(
+    (menuItems) => {
+      if (isPivotingAvailable) {
+        return [...menuItems, 'columnMenuPivotItem'];
+      }
+      return menuItems;
+    },
+    [isPivotingAvailable],
+  );
+
+  useGridRegisterPipeProcessor(apiRef, 'columnMenu', addColumnMenuButton);
+
+  useGridApiMethod(apiRef, { setPivotModel, setPivotActive, setPivotPanelOpen }, 'public');
```

Comment:
Should we also nest the pivoting API methods, like data source?

See https://github.com/mui/mui-x/pull/16992#discussion_r2014847280